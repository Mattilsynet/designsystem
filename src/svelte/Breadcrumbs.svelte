<script lang="ts">
  import {onMount} from 'svelte';
  import type {Breadcrumbs} from '../ts/types';
  import {useMachine} from '@xstate/svelte';
  import {createMachine} from 'xstate';

  const BUTTON_ELLIPSIS = Symbol();
  export let breadcrumbs: Breadcrumbs = {items: []};
  export let loadJs = false;
  const LIMIT_BEFORE_PARTIAL = 3;

  interface BreadcrumbsContext {
    breadcrumbsItems: Array<{url: string; title: string} | symbol>;
  }

  type BreadcrumbsEvent = {type: 'MOUNTED'} | {type: 'TOGGLE'};

  type BreadcrumbsState =
    | {value: 'serverRendered'; context: BreadcrumbsContext}
    | {value: 'partial'; context: BreadcrumbsContext}
    | {value: 'full'; context: BreadcrumbsContext};

  const breadcrumbsMachine = createMachine<BreadcrumbsContext, BreadcrumbsEvent, BreadcrumbsState>(
    {
      id: 'breadcrumbs',
      initial: 'serverRendered',
      context: {
        breadcrumbsItems: breadcrumbs.items
      },
      states: {
        serverRendered: {
          on: {
            MOUNTED: [
              {
                target: 'partial',
                cond: context => context.breadcrumbsItems.length > LIMIT_BEFORE_PARTIAL
              },
              {
                target: 'full'
              }
            ]
          }
        },
        partial: {
          entry: 'prepareCollapsedBreadcrumbs',
          on: {TOGGLE: 'full'}
        },
        full: {
          entry: 'expandBreadcrumbs',
          on: {TOGGLE: 'partial'}
        }
      }
    },
    {
      actions: {
        prepareCollapsedBreadcrumbs: context => {
          context.breadcrumbsItems = [
            breadcrumbs.items[0],
            breadcrumbs.items[1],
            BUTTON_ELLIPSIS,
            breadcrumbs.items[breadcrumbs.items.length - 1]
          ];
        },
        expandBreadcrumbs: context => {
          context.breadcrumbsItems = breadcrumbs.items;
        }
      }
    }
  );

  const {state, send} = useMachine(breadcrumbsMachine);

  let ariaLabel = breadcrumbs.ariaLabel ?? 'brÃ¸dsmulesti';
  let showAllBreadCrumbsLabel = breadcrumbs.showAllAriaLabel ?? 'vis mer';
  let homeLabel = breadcrumbs.homeLabel ?? 'Hjem';

  $: isFull = $state.value !== 'partial';
  $: onServer = $state.value === 'serverRendered';
  $: preparedBreadCrumbs = $state.context.breadcrumbsItems;

  if (loadJs) {
    onMount(() => send('MOUNTED'));
  }
</script>

<nav class="breadcrumbs" aria-label={ariaLabel}>
  <ol class:expanded={isFull}>
    {#each preparedBreadCrumbs as item, index}
      <li class:ellipsis={!isFull && index + 1 === preparedBreadCrumbs.length}>
        {#if index === 0}
          <a href={item.url}>{homeLabel}</a>
        {:else if item === BUTTON_ELLIPSIS}
          <button
            type="button"
            aria-expanded="false"
            aria-label={showAllBreadCrumbsLabel}
            class="button button--link forward-arrow"
            on:click={() => send('TOGGLE')}
          >
            ...
          </button>
        {:else if index + 1 < preparedBreadCrumbs.length}
          <a class="forward-arrow" href={item.url}>{item.title}</a>
        {:else}
          <a class="forward-arrow" aria-current="page" href={item.url}>{item.title}</a>
        {/if}
      </li>
    {/each}
  </ol>
</nav>
