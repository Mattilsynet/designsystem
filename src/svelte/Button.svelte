<script lang="ts">
  export let progressIndicator = false
  let inProgress = false
  let formInProgressLabel = ''

  function setInProgress(e){
    if(progressIndicator && inProgress){
      e.preventDefault()
    } else {
      inProgress = true
      e.target.blur()
      formInProgressLabel = 'Sender inn skjema, venter p√• svar'
      e.target.focus()
    }
  }
</script>

<button on:click={setInProgress} class="button button--primary self-start" aria-label={formInProgressLabel} >
  <slot/>
  {#if progressIndicator && inProgress}
    <span class="spinner"/>
  {/if}
</button>
