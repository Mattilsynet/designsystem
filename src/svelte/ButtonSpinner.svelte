<script lang="ts">
  export let formInProgressAriaLabel = ''
  export let color: 'primary' | 'secondary' = 'primary'
  export let spinnerPlacement: 'start' | 'end' = 'end'

  let inProgress = false
  let formInProgressLabel = ''
  let colorClass = color === 'primary' ? 'button--primary' : 'button--secondary'

  function setInProgress(e){
    if(inProgress){
      e.preventDefault()
    } else {
      inProgress = true
      e.target.blur()
      formInProgressLabel = formInProgressAriaLabel
      e.target.focus()
    }
  }
</script>

<button on:click={setInProgress} class="button button--spinner {colorClass}" aria-label={formInProgressLabel} >
  {#if inProgress && spinnerPlacement === 'start'}
    <span class="spinner spinner--start"/>
  {/if}
  <slot/>
  {#if inProgress && spinnerPlacement === 'end'}
    <span class="spinner spinner--end"/>
  {/if}
</button>
