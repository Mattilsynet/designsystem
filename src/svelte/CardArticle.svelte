<script lang="ts">
  export let headingId: string
  export let title: string
  export let intro: string | undefined
  export let text: string | undefined
  export let linkUrl: string | undefined
  export let linkText: string | undefined
  export let type: 'task' | 'form-intro-module' | 'legal-text' | '' = ''
  export let headerTag: 'h2' | 'h4' | 'h3' = 'h3' // Fallback is h3
</script>

<article class={type} aria-labelledby={headingId}>
  {#if headerTag === 'h2'}
    <h2 id={headingId} class="title h3">{@html title}</h2>
  {:else if headerTag === 'h4'}
    <h4 id={headingId} class="title h4">{@html title}</h4>
  {:else}
    <h3 id={headingId} class="title">{@html title}</h3>
  {/if}

  {#if intro}
    <div class="intro">
      {@html intro}
    </div>
  {/if}

  {#if text}
    <div class="text">
      {@html text}
    </div>
  {/if}

  {#if linkUrl && linkText}
    <a href={linkUrl} class="button button--primary self-start">
      {linkText}
    </a>
  {/if}
</article>
