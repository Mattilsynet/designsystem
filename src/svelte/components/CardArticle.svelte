<script lang="ts">
  import HeadingLevel from './HeadingLevel.svelte'
  import {mapRelExternal} from '../../ts/utils'

  export let headingId: string
  export let title: string
  export let intro: string | undefined
  export let text: string | undefined
  export let linkUrl: string | undefined
  export let linkText: string | undefined
  export let type: 'task' | 'form-intro-module' | 'legal-text' | '' = ''
  export let headerTag: 'h2' | 'h3' | 'h4' | 'h5' | 'h6' = 'h3' // Fallback is h3
</script>

<article class={type} aria-labelledby={headingId}>
  <HeadingLevel
    id={headingId}
    class="title {headerTag === 'h2' ? 'h3' : ''}"
    headingLevel={+headerTag.charAt(1)}>
    {@html title}
  </HeadingLevel>

  {#if intro}
    <div class="intro">
      {@html intro}
    </div>
  {/if}

  {#if text}
    <div class="text">
      {@html text}
    </div>
  {/if}

  {#if linkUrl && linkText}
    <a href={linkUrl} rel={mapRelExternal(linkUrl)} class="button button--primary self-start">
      {linkText}
    </a>
  {/if}
</article>
