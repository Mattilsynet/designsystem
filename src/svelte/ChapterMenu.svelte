<script lang="ts">
  import {createEventDispatcher} from 'svelte'
  import type {Chapter, ChapterChangeDetails} from '../ts/types'

  export let chapters: Array<Chapter>
  export let currentChapterNumber = 0
  export let menuTitle: string
  export let showChapterNumbers = false

  const dispatch = createEventDispatcher<{chapterChange: ChapterChangeDetails}>()
</script>

<nav class="chapter-menu" aria-labelledby="chapter-menu-title">
  <h3 id="chapter-menu-title" class="h4">{menuTitle}</h3>

  <ol class="list-unstyled">
    {#each chapters as chapter, index}
      <li>
        <a
          on:click|preventDefault={dispatch('chapterChange', {index: chapter.index})}
          href={chapter.url}
          aria-current={chapter.index === currentChapterNumber ? 'page' : undefined}>
          {#if showChapterNumbers}
            {index + 1}
          {/if}
          {chapter.heading}
        </a>
      </li>
    {/each}
  </ol>
</nav>
