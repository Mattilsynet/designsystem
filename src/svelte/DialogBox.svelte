<script context="module" lang="ts">
  let instanceCounter = 0
</script>

<script lang="ts">
  import {fade} from 'svelte/transition'
  export let isOpen: boolean = true
  export let title: string = ''
  export let ariaTitle: string = ''
  export let closeBtnAriaLabel: string = 'Lukk'
  const dialogBoxHeadingId = `ui-dialog-box-${instanceCounter++}`
  const dialogCloseButtonId = `dialog-close-button-${instanceCounter++}`

  function handleClose() {
    isOpen = false
  }
</script>

{#if isOpen}
  <div
    class={isOpen ? 'dialog-box' : 'dialog-box--closed'}
    data-testid="dialog-box"
    role="dialog"
    tabindex="0"
    transition:fade
    aria-hidden={!isOpen}
    aria-labelledby={dialogBoxHeadingId}>
    <h2 id={dialogBoxHeadingId} class="h4 dialog-box--title {title ? '' : 'inclusively-hidden'}">
      {title ? title : ariaTitle}
    </h2>
    <button
      type="button"
      class="button button--link dialog-box--close-button "
      aria-label={closeBtnAriaLabel} />

    <div class="dialog-box--content">
      <slot />
    </div>
  </div>
{/if}
