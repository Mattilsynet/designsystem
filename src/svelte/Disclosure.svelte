<script lang="ts" context="module">
  let counter = 0;
</script>

<script lang="ts">
  import {onMount} from 'svelte';
  import {slide} from 'svelte/transition';
  import {useMachine} from '@xstate/svelte';
  import {createMachine, assign} from 'xstate';

  export let loadJs = true;
  export let title: string;
  export let headerTag: 'h2' | 'h3' = 'h3';

  interface DisclosureContext {
    isOpen: boolean;
    isFirstRenderFinished: boolean;
  }

  type DisclosureEvent = {type: 'MOUNTED'} | {type: 'TOGGLE'};

  type DisclosureState =
    | {value: 'serverRendered'; context: DisclosureContext}
    | {value: 'closed'; context: DisclosureContext}
    | {value: 'open'; context: DisclosureContext};

  const bodyId = `ui-disclosure-${counter++}`;
  const disclosureMachine = createMachine<DisclosureContext, DisclosureEvent, DisclosureState>({
    id: 'disclosure',
    initial: 'serverRendered',
    context: {
      isOpen: true,
      isFirstRenderFinished: false
    },
    states: {
      serverRendered: {
        on: {MOUNTED: 'closed'}
      },
      closed: {
        entry: assign({isOpen: false}),
        exit: assign({isFirstRenderFinished: true}),
        on: {TOGGLE: 'open'}
      },
      open: {
        entry: assign({isOpen: true}),
        on: {TOGGLE: 'closed'}
      }
    }
  });
  const {state, send} = useMachine(disclosureMachine);

  $: isOpen = $state.context.isOpen;
  $: onServer = $state.value === 'serverRendered';

  if (loadJs) {
    onMount(() => send('MOUNTED'));
  }
</script>

<div class="disclosure">
  {#if onServer && headerTag === 'h2'}
    <h2 class="disclosure-header">{title}</h2>
  {:else if onServer}
    <h3 class="disclosure-header">{title}</h3>
  {:else}
    <button
      class="disclosure-header {headerTag === 'h2' ? 'h2' : ''}"
      aria-expanded={isOpen}
      aria-controls={bodyId}
      on:click={() => send('TOGGLE')}
    >
      {title}
    </button>
  {/if}

  {#if isOpen}
    <div
      id={bodyId}
      class="disclosure-panel"
      transition:slide|local={{duration: $state.context.isFirstRenderFinished ? 300 : 0}}
    >
      {#if !onServer && headerTag === 'h2'}
        <h2 class="inclusively-hidden">{title}</h2>
      {:else if !onServer}
        <h3 class="inclusively-hidden">{title}</h3>
      {/if}
      <slot />
    </div>
  {/if}
</div>
