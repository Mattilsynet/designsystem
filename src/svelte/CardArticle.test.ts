/**
 * @jest-environment jsdom
 */

import {getByTestId, render} from '@testing-library/svelte'
import CardArticle from './CardArticle.svelte'

describe('Card Article', () => {
  test('Renders', () => {
    const componentOptions = {
      headingId: 'testId',
      title: 'Varsle om piggtrådgjerder',
      intro: '',
      text: `<p>Du bør varsle Mattilsynet hvis du oppdager</p>
           <ul>
             <li>at noen setter opp nye piggtrådgjerder</li>
             <li>eldre piggtrådgjerder som utgjør en risiko for at dyr blir skadet</li>
           </ul>`,
      formLinkUrl: 'http://',
      formLinkText: 'Lenke til skjema',
      disableCss: false
    }

    const {getByText} = render(CardArticle, {props: componentOptions})
    expect(getByText('Varsle om piggtrådgjerder')).toBeInTheDocument()
    expect(getByText('Varsle om piggtrådgjerder')).toMatchSnapshot(
      '<h3 class="title" id="testId">Varsle om piggtrådgjerder</h3>'
    )
  })
  test('Renders with h2 heading', () => {
    const componentOptions = {
      headingId: 'testId',
      title: 'Varsle om piggtrådgjerder',
      intro: '',
      text: `<p>Du bør varsle Mattilsynet hvis du oppdager</p>
           <ul>
             <li>at noen setter opp nye piggtrådgjerder</li>
             <li>eldre piggtrådgjerder som utgjør en risiko for at dyr blir skadet</li>
           </ul>`,
      formLinkUrl: 'http://',
      formLinkText: 'Lenke til skjema',
      disableCss: false,
      headerTag: 'h2'
    }

    const {getByText} = render(CardArticle, {props: componentOptions})
    expect(getByText('Varsle om piggtrådgjerder')).toBeInTheDocument()
    expect(getByText('Varsle om piggtrådgjerder')).toMatchSnapshot(
      '<h2 class="title" id="testId">Varsle om piggtrådgjerder</h2>'
    )
  })
})
