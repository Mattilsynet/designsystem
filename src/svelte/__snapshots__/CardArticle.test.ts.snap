// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Article Renders with h2 heading: <h2 class="title" id="testId">Varsle om piggtr책dgjerder</h2> 1`] = `
<h2
  class="title h3"
  id="testId"
>
  Varsle om piggtr책dgjerder
</h2>
`;

exports[`Card Article Renders: <h3 class="title" id="testId">Varsle om piggtr책dgjerder</h3> 1`] = `
<h3
  class="title"
  id="testId"
>
  Varsle om piggtr책dgjerder
</h3>
`;
